lib/index.js

```diff
@@ -36,143 +64,115 @@ module.exports = {
-          cb();
-        });
-        server.on('InternalServer', function(req, res, err, cb) {
-          res.send(500, { type: 'ServerError', message: err.message });
-          cb();
-        });
-        server.on('uncaughtException', function(req, res, route, err) {
-          logger(req, res, route, err);
-          if (!res.headersSent) {
-            res.send(err.statusCode || 500, { type: err.type || 'Error', message: err.message });
-          }
-        });
+      });
+      routes[api.basePath] = [{
+        name: 'index',
+        method: 'get',
+        version: api.info.version,
+        handler: indexRouteHandler(api)
+      }];
+    });
 
-        server.listen(
-          config.get('port'),
-          config.get('host'),
-          function() {
-            debug('%s listening at %s', server.name, server.url);
-          }
-        );
 
-        return server;
+    const corsMap = {};
+    keys(routes).sort().reverse().forEach((routePath) => {
+      corsMap[routePath] = { shouldSet: true, expectedHeaders: [], allowedMethods: ['OPTIONS'] };
+      routes[routePath].forEach((action) => {
+        if (action.method === 'opts') { corsMap[routePath].shouldSet = false; }
+        corsMap[routePath].allowedMethods.push(action.method.toUpperCase());
+        corsMap[routePath].expectedHeaders.concat(action.expectedHeaders);
 
-      }).catch(function(err) {
-        debug(err);
-        throw err;
+        debug('routing %s@%s for %s', routePath, action.version, action.name);
+        server[action.method](
+          { path: routePath, version: action.version },
+          cors.headers(),
+          action.handler
+        );
       });
+    });
+
+    keys(corsMap).forEach((route) => {
+      cors.preflight({ path: route, ...corsMap[route] });
+    });
+
+    server.on('after', function(req, res, route, err) {
+      logger(req, res, route, err);
+    });
+    server.on('NotFound', function(req, res, err, cb) {
+      res.send(404, { type: 'NotFound', message: err.message });
+      cb();
+    });
+    server.on('MethodNotAllowed', function(req, res, err, cb) {
+      res.send(405, { type: 'MethodNotAllowed', message: err.message });
+      cb();
+    });
+    server.on('VersionNotAllowed', function(req, res, err, cb) {
+      res.send(400, { type: 'VersionNotAllowed', message: err.message });
+      cb();
+    });
+    server.on('InternalServer', function(req, res, err, cb) {
+      res.send(500, { type: 'ServerError', message: err.message });
+      cb();
+    });
+    server.on('uncaughtException', function(req, res, route, err) {
+      logger(req, res, route, err);
+      if (!res.headersSent) {
+        res.send(err.statusCode || 500, { type: err.type || 'Error', message: err.message });
+      }
+    });
+
+    const listenDefer = defer();
+    server.listen(config.get('port'), config.get('host'), () => {
+      debug('%s listening at %s', server.name, server.url);
+      listenDefer.resolve();
```

Comment:
can this callback have an error? i.e., resolve vs reject and all that

Reply:
as far as I can tell, the callback doesn't receive any data. I'm not sure how we'd decide to reject?